### PR TITLE
oauth2: preserve query parameters in the authorization_endpoint

### PR DESCRIPTION
**Pull request by qfiard**
Query parameters are needed to customize the authentication flow for some providers, e.g. for Google Oauth2
- access_type=offline is needed to generate both an access and refresh token
- prompt=select_account is needed to ask the user to select an account
https://developers.google.com/identity/protocols/oauth2/openid-connect#authenticationuriparameters

This could be supported as fields in the envoy.extensions.filters.http.oauth2.v3.OAuth2 proto but it feels simpler to preserve URL parameters until this is needed by more clients.

Signed-off-by: Quentin Fiard <quentin.fiard@gmail.com>

### CODE REVIEW

**Comment by rgs1** [comment]
@qfiard any chance you could merge main and a release note so we can land this?

**Comment by rgs1** [comment]
(If not, I'll pick it up -- I think it's good to have)

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (qfiard) [reply test/extensions/filters/http/oauth2/filter_test.cc:330]: Done.

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (qfiard) [reply test/extensions/filters/http/oauth2/filter_test.cc:343]: Done.

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (qfiard) [reply test/extensions/filters/http/oauth2/filter_test.cc:347]: Done.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (qfiard) [reply source/extensions/filters/http/oauth2/filter.cc]: Added a check in the constructor to make sure the authorization endpoint is a valid URL (consistent with how oauth_token_endpoint_.cluster() is checked) and an ASSERT here. Seems preferable to returning sendUnauthorizedResponse() as an error is returned when envoy is started rather than on the first oauth request, but please let me know if you have any concerns.

**Comment by qfiard** [comment]
Done, apologies for the delay and thanks for the feedback. Should be ready for a second pass.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (rgs1) [inline source/extensions/filters/http/oauth2/filter.cc:342]: not a huge deal, but we could save this map (`query_params`) in the constructor and just make a copy of the map on every request, should be cheaper than parsing the authorization endpoint on every request. Maybe leave a TODO if you don't think we should handle it now?

**Review by rgs1 — COMMENTED** [review]
Thanks for the update -- LGTM modulo one final ask/nit.

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (qfiard) [inline test/extensions/filters/http/oauth2/filter_test.cc:315]: Removed since it is tested through the location header below and would otherwise require keeping the encodedAuthScopes() method solely for testing purposes.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (qfiard) [reply source/extensions/filters/http/oauth2/filter.cc]: Done, good feedback. I am also now caching the URL object to avoid parsing the URL on every object, which has the additional benefit of avoiding the ASSERT when using URL parts. Please let me know if you have any concerns.

**Comment by qfiard** [comment]
Should be ready for another pass. Thanks for the review.

**Comment by rgs1** [comment]
@qfiard thanks! Mind addressing the formatting issue please?

**Comment by rgs1** [comment]
it's this:

```
INFO: Running command line: bazel-bin/tools/docs/rst_check
RSTChecker NOTICE [current_version] Running check
RSTChecker ERROR [current_version] (docs/root/version_history/current.rst:28) Version history line malformed. Does not match VERSION_HISTORY_NEW_LINE_REGEX
 * oauth2: preserve query parameters in the :ref:`authorization_endpoint <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.authorization_endpoint>`.
Please use messages in the form 'category: feature explanation.', starting with a lower-cased letter and ending with a period.
RSTChecker ERROR [current_version] Check failed
RSTChecker ERROR Summary

```

**Comment by qfiard** [comment]
Done. Thanks for the review!

**Review by snowp — DISMISSED** [review]
Thanks, LGTM!

/retest

**Comment by snowp** [comment]
Seems like this PR brings the coverage for the oauth code below the threshold, can you take a look?

```
Code coverage for source/extensions/filters/http/oauth2 is lower than limit of 96.6 (96.3)
```

/wait

**Comment by qfiard** [comment]
I added a test for the case where the authorization endpoint is invalid, which as far as I can tell was the main case not covered by tests: https://storage.googleapis.com/envoy-pr/27bd365/coverage/source/extensions/filters/http/oauth2/filter.cc.gcov.html
Could you rerun the presubmit? Thanks for catching!

**Comment by alyssawilk** [comment]
I think your reviewers were waiting on CI passing, but this now needs a main merge
/wait

**Comment by qfiard** [comment]
Merged with main, apologies for the delay. Should be ready for another review.

**Pull request closed without merging**
